File: 0042-Trapping-Rain-Water/lc0042.cc
// C++ solution.

// Similar to question 11. We can use two pointers method.
// Instead of calculating area by height*width, we can think it in a cumulative way.
// Search from left to right and maintain a max height of left and right separately,
// which is like a one-side wall of partial container.
// If current height of left is lower, we fill water in the left bin, and vice versa.
// Until left meets right, we filled the whole container.
class Solution {
public:
    int trap(vector<int>& height) {
        int posl = 0;
        int posr = height.size()-1;
        int left_max = 0;
        int right_max = 0;
        int ans = 0;
        while (posl < posr) {
            if (height[posl] < height[posr]) {
                left_max = std::max(left_max, height[posl]);
                ans += left_max - height[posl++];
            } else {
                right_max = std::max(right_max, height[posr]);
                ans += right_max - height[posr--];
            }
        }
        return ans;
    }
};

// Just use monostack.
// Use a stack to preserve a descending sequence.
// For example: [3,2,1,0,3]
// *---*
// **--*
// ***-*
class Solution {
public:
    int trap(vector<int>& height) {
        std::stack<int> stk;
        int ans = 0;
        for (int i = 0; i < height.size(); i++) {
            for (;!stk.empty() && height[stk.top()] < height[i];) {
                int barh = height[stk.top()];
                stk.pop();
                if (!stk.empty()) {
                    ans += (std::min(height[stk.top()], height[i])-barh) * (i-stk.top()-1);
                }
            }
            stk.push(i);
        }
        return ans;
    }
};

File: 0042-Trapping-Rain-Water/lc0042.rs
// Rust solutions.

// O(n) space.
impl Solution {
    pub fn trap(height: Vec<i32>) -> i32 {
        if height.is_empty() { return 0; }
        let sz = height.len();
        let mut leftmax: Vec<i32> = vec![height[0]; sz];
        let mut rightmax: Vec<i32> = vec![height[sz-1]; sz];

        for i in 1..sz {
            leftmax[i] = std::cmp::max(height[i], leftmax[i-1]);
        }
        for i in (0..sz-1).rev() {
            rightmax[i] = std::cmp::max(height[i], rightmax[i+1]);
        }

        let mut ans = 0;
        for i in 0..sz {
            ans += std::cmp::min(leftmax[i], rightmax[i]) - height[i];
        }
        ans
    }
}

// O(1) space.
impl Solution {
    pub fn trap(height: Vec<i32>) -> i32 {
        if height.is_empty() { return 0; }
        let (mut left, mut right) = (0_usize, height.len()-1);
        let (mut leftmax, mut rightmax) = (0_i32, 0_i32);
        let mut ans = 0;
        while left < right {
            if height[left] < height[right] {
                leftmax = std::cmp::max(leftmax, height[left]);
                ans += leftmax - height[left];
                left += 1;
            } else {
                rightmax = std::cmp::max(rightmax, height[right]);
                ans += rightmax - height[right];
                right -= 1;
            }
        }
        ans
    }
}
